### ISSUE TITLE
Model Accuracy Lower Than Reported in Paper – Seeking Insights

### ISSUE THREAD

**Issue by Hurwitzzz** [malcolmsailor/rnbert#1]
## **Problem Description**  

First of all, thank you for your excellent work!  

I trained the unconditioned Roman numeral model following the instructions in the README. The loss and accuracy curves during training appear reasonable—the loss decreases rapidly within the first 1,000 iterations and then gradually declines until the end of training. The accuracy curve follows a similar pattern. Below are the corresponding plots:

<details>
  <summary>Click here to show/hide training plots</summary>

  ![Image](https://github.com/user-attachments/assets/930f13af-c5e4-4f98-be7f-235daa1aae79)  

  ![Image](https://github.com/user-attachments/assets/211cd4d8-18a1-4594-9c83-7de72ccdd56a)

</details>

**However**, the final performance does not match the reported accuracy in the paper. I evaluated the model using `rnbert_unconditioned_metrics.sh` and obtained the following results:  

<details>
  <summary>Click here to show/hide training scripts and results</summary>

  ```bash
  # Training script
  python musicbert_fork/training_scripts/train_chord_tones.py \
      -a base \
      -d ${RNDATA_ROOT}/rnbert_rn_uncond_data_bin \
      --multitask \
      --validate-interval-updates 2500 \
      --lr 0.00025 \
      --batch-size 4 \
      --fp16 \
      --freeze-layers 9 \
      --total-updates 50000 \
      --warmup-updates 2500 \
  
  # Corresponding results on the test set
  ,accuracy,balanced_accuracy
  key_pc,0.75094,0.67481
  mode,0.80952,0.78937
  KEY,0.67324,0.56704
  DEGREE,0.50738,0.019606
  QUALITY,0.54238,0.14274
  INVERSION,0.66523,0.2538
  primary_alteration_primary_degree_secondary_alteration_secondary_degree,0.50738,0.019606
  quality,0.54238,0.14274
  inversion,0.66523,0.2538
  DQI,0.2809,0.0062941
  key_pc,0.75094,0.67481
  mode,0.80952,0.78937
  primary_alteration_primary_degree_secondary_alteration_secondary_degree,0.50738,0.019606
  quality,0.54238,0.14274
  inversion,0.66523,0.2538
  KDQI,0.23404,0.019206
  ```

</details>

Then, I observed that **increasing the batch size improved the model's performance**. Below are the best results I achieved so far, which are closer to the reported accuracy but still have a gap.

<details>
  <summary>Click here to show/hide training scripts and results</summary>

  ```bash
  # Training script. Trained on one Nvidia A40 GPU
  python musicbert_fork/training_scripts/train_chord_tones.py \
      -a base \
      -d ${RNDATA_ROOT}/rnbert_rn_uncond_data_bin \
      --multitask \
      --validate-interval-updates 2500 \
      --lr 0.0003 \
      --batch-size 32 \
      --fp16 \
      --freeze-layers 9 \
      --total-updates 50000 \
      --warmup-updates 2500 \
  
  # Corresponding results on the test set
  ,accuracy,balanced_accuracy
  key_pc,0.8178,0.80655
  mode,0.85108,0.83291
  KEY,0.75095,0.75364
  DEGREE,0.68541,0.12674
  QUALITY,0.81226,0.44021
  INVERSION,0.81388,0.62987
  primary_alteration_primary_degree_secondary_alteration_secondary_degree,0.68541,0.12674
  quality,0.81226,0.44021
  inversion,0.81388,0.62987
  DQI,0.55205,0.099712
  KDQI,0.49869,0.20904
```
</details>

## **Questions**

1. Are there any potential issues in my training setup that might have caused the performance gap?
2. Did you use multiple GPUs for training?
      -   If so, how many GPUs were used?
      -   Since Fairseq defaults to using all available GPUs, knowing the total batch size (batch size per GPU * number of GPUs) would help in understanding the difference in training dynamics.

Any insights would be greatly appreciated. Thank you!

**Comment by malcolmsailor** [comment]
Thanks so much for this issue. In response to your question, I only used 1 GPU for training.

I'll have to look further into this and get back to you. Just to be sure, are you fine-tuning the musicbert checkpoint? Or training from scratch?

**Comment by malcolmsailor** [comment]
So I installed new environments, re-cloned the repository, re-built the data, and am running the fine-tuning, all according to the directions in the README. The training has only begun but it already looks like I will be unable to reproduce your behavior. For example, the very first validation loss that I get during training is 1.117, which appears to be lower than the best validation loss from your training run. (This is with the original batch size.)

**Comment by malcolmsailor** [comment]
For now the first thing is to make sure that you are loading the correct MusicBERT checkpoint for fine tuning. Just to double check, are you downloading the checkpoint from https://1drv.ms/u/s!Av1IXjAYTqPsuBaM9pZB47xjX_b0?e=wg2D5O and then setting MUSICBERT_DEFAULT_CHECKPOINT to its path?

**Comment by Hurwitzzz** [comment]
### Problem Solved!
Thank you so much for your quick response!

Your assumption was absolutely correct—such strong problem-solving skills! The issue was indeed related to checkpoint loading. I had a typo in the path of the environment variable, which caused the pretrained checkpoint not to load correctly.

Although the training has not yet finished, the current results have already significantly outperformed my previous training runs and are approaching the accuracy reported in the paper.

I will now close this issue. Once again, thank you for your fast and effective support!

**Comment by malcolmsailor** [comment]
I'm glad that solved it! Let me know if you have any other issues.

BTW I just published a [new repository](https://github.com/malcolmsailor/musicbert_hf) that ports much of this code to HuggingFace. If you're planning to use this architecture for anything, it could be useful to you, since HuggingFace is a good deal easier to use than FairSeq.

**Comment by Hurwitzzz** [comment]
> I'm glad that solved it! Let me know if you have any other issues.
> 
> BTW I just published a [new repository](https://github.com/malcolmsailor/musicbert_hf) that ports much of this code to HuggingFace. If you're planning to use this architecture for anything, it could be useful to you, since HuggingFace is a good deal easier to use than FairSeq.

Great! If I get the chance, I’d love to contribute to it as well.

**Comment by malcolmsailor** [comment]
> If I get the chance, I’d love to contribute to it as well.

Awesome! Even if you get a chance to go through the readme and see if you can run the examples, that would be helpful.
